net: context: Fix the ICMP error on udp

sasataku opened the pull request:
This commit fixes the ##70020.

When receiving a UDP packet, net_conn_input() searches for a matching connection within `conn_used`.

However, when receiving UDP packets simultaneously from multiple clients, we may encounter a situation where the connection that was supposed to be bound cannot be found within `conn_used`, and raise the ICMP error.

This is because, within recv_udp(), to avoid the failure of bind_default(), we temporarily remove it from `conn_used` using net_conn_unregister().

If the context already has a connection handler, it means it's already registered. In that case, all we have to do is 1) update the callback registered in the net_context and 2) update the user_data using net_conn_change_callback().

Since net_conn_change_callback() takes the _other_ callback function, net_context_packet_received() (I know it's confusing), this must be the same function passed in net_conn_register() down below.

-----

Fixes #70020

sasataku committed:
net: context: Fix the ICMP error on udp

When receiving a UDP packet, net_conn_input() searches for a
matching connection within `conn_used`.

However, when receiving UDP packets simultaneously from multiple
clients, we may encounter a situation where the connection that was
supposed to be bound cannot be found within `conn_used`, and raise
the ICMP error.

This is because, within recv_udp(), to avoid the failure of
bind_default(), we temporarily remove it from `conn_used` using
net_conn_unregister().

If the context already has a connection handler, it means it's
already registered. In that case, all we have to do is 1) update
the callback registered in the net_context and 2) update the
user_data and remote address and port using net_conn_update().

Fixes #70020

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku commented:
I've tested the some zephyr test that called net_context_recv() as below, and all passed.

- tests/net/context
   ```
   SUITE PASS - 100.00% [net_context]: pass = 15, fail = 0, skip = 0, total = 15 duration = 5.015 seconds
   ```
- tests/net/ipv6
  ```
  SUITE PASS - 100.00% [net_ipv6]: pass = 26, fail = 0, skip = 0, total = 26 duration = 2.456 seconds
   ```
- tests/net/tcp
   ```
   SUITE PASS - 100.00% [net_tcp]: pass = 18, fail = 0, skip = 0, total = 18 duration = 6.268 seconds
   ```
-  tests/net/traffic_class
   ```
   SUITE PASS - 100.00% [net_traffic_class]: pass = 11, fail = 0, skip = 0, total = 11 duration = 0.183 seconds
   ```
- tests/net/virtual/
   ```
   SUITE PASS - 100.00% [net_virtual]: pass = 10, fail = 0, skip = 0, total = 10 duration = 0.010 seconds
   ```

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2401,9 +2401,21 @@ static int recv_udp(struct net_context *context,
+	 * the callback registered in the net_context and 2) update the
+	 * user_data using net_conn_change_callback().
+	 *
+	 * Since net_conn_change_callback() takes the _other_ callback
+	 * function, net_context_packet_received() (I know it's confusing),
+	 * this must be the same function passed in net_conn_register() down below.
```
This comment needs some clarification, it is a bit difficult to follow, could you formulate it better. Also remove the `(I know it's confusing)` sentence.

sasataku replied:
Okay. I will update the comment.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2401,9 +2401,21 @@ static int recv_udp(struct net_context *context,
-		context->conn_handler = NULL;
+		context->recv_cb = cb;
+		ret = net_conn_change_callback(context->conn_handler,
+					 net_context_packet_received,
+					 user_data);
```
Please fix the indentation here, these two lines should start the same column as the `context->conn_handler`

sasataku replied:
Sorry for missing it. I will fix the indentation.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2401,9 +2401,21 @@ static int recv_udp(struct net_context *context,
 
```
Could you change the `This commit fixes the #70020.` from the commit message into `Fixes #70020` and place it before the signed-off-by line. This way it is easier to grep the log later if one needs to figure out the exact commit that fixed it. See examples in other commits in the tree

sasataku replied:
Okay, I will fix it.

jukkar reviewed (commented):
Looks good. Could you also fix the `recv_raw()` which has the same issue? Just add separate commit for that in this PR.

sasataku commented:
> Some of the network tests did not like this change, could you check the CI logs?

This commit results in the failure of `test_dgram_connected()`.
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/udp/src/main.c#L1307

I suspect this is because the remote address set by the `connect()` after the `bind()` is not being reflected in the connection.
I'm currently considering a corrective approach, and I believe we need a function to update the remote address and flags set only within `net_conn_register()`.

sasataku committed:
net: conn: Move net_conn_change_callback() to static

The net_conn_change_callback() is not currently being called by
anyone, so this commit moves to static function, and replaces
the change callback parameter process in net_conn_register().

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku committed:
net: context: Fix the ICMP error on raw

This commit applies the issues detected in UDP to recv_raw() as
well. Please refer to the previous commit log for details.

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku commented:
> This commit results in the failure of test_dgram_connected().

I've updated the PR for resolve the  failure of `test_dgram_connected()`.
However I had other unit test failure, so I will investigate it.

yashi commented on subsys/net/ip/net_context.c:
```diff
@@ -2401,9 +2401,29 @@ static int recv_udp(struct net_context *context,
-		net_conn_unregister(context->conn_handler);
-		context->conn_handler = NULL;
+		context->recv_cb = cb;
+		ret = net_conn_change_callback(context->conn_handler,
+					       net_context_packet_received,
+					       user_data);
+		if (ret) {
+			goto end;
+		}
+		ret = net_conn_change_remote(context->conn_handler,
+					     context->flags & NET_CONTEXT_REMOTE_ADDR_SET ?
+								&context->remote : NULL,
+					     ntohs(net_sin(&context->remote)->sin_port));
```
How about having `net_conn_update()` and do both.  Because we don't have any user of `net_conn_change_callback()` in tree and it's not a public API anyway.

So that a call-site will become something like:

```c
if (context->conn_handler) {
    ret = net_conn_update();
} else {
    ret = bind_default();
    if (ret >= 0) {
        ret = net_conn_register();
    }
}

return ret;
```

@jukkar any thoughts?

yashi replied:
I think having a static version of `net_conn_change_callback()` and `net_conn_change_remote()` in connection.c is fine. I assume that it's better to have `net_conn_update()` for net_context.c to consume.

jukkar replied:
@yashi good point, I like your `net_conn_update()` idea, we could do it like that.

sasataku replied:
Thank you for your review.
I will update the PR based on `net_conn_update()` idea.

sasataku commented:
I've passed the all unit tests.

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -280,6 +280,7 @@ int net_conn_register(uint16_t proto, uint8_t family,
 {
```
Minor nit, but you should put the `Fixes #70020` to only one commit that has the actual fix. It looks like it is the 2nd one, all the other commits do not have to mention that.
So in the first commit, you can remove the
```
This is a prepared commit for fixes https://github.com/zephyrproject-rtos/zephyr/issues/70020.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/70020
```
lines.

sasataku replied:
Thank you. I will update the commit log.

sasataku committed:
net: conn: Add static function for changing remote

This commit adds the new static function for change the remote
address and port to connection, and replaces the changing process
for remote address and port in net_conn_register().

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

sasataku committed:
net: conn: Add internal function for update connection

This commit adds the new internal function for update the callback,
user data, remote address, and port for a registered connection
handle.

Signed-off-by: Takuya Sasaki <takuya.sasaki@spacecubics.com>

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2401,53 +2401,68 @@ static int recv_udp(struct net_context *context,
+				      context->flags & NET_CONTEXT_REMOTE_ADDR_SET ?
+						&context->remote : NULL,
+				      ntohs(net_sin(&context->remote)->sin_port));
+	} else {
 
```
The else branch is certainly one option, it is just that there is lot of code to indent which makes the lines longer. We could avoid the else branch like it was originally. For example like this:
```
if (context->conn_handler != NULL) {
	context->recv_cb = cb;
	ret = net_conn_update(context->conn_handler,
			      net_context_packet_received,
			      user_data,
			      context->flags & NET_CONTEXT_REMOTE_ADDR_SET ?
					&context->remote : NULL,
			      ntohs(net_sin(&context->remote)->sin_port));
	return ret;
}
```

sasataku replied:
Okay, fixed.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2512,25 +2512,38 @@ static int recv_raw(struct net_context *context,
-	if (ret) {
-		return ret;
-	}
+		ret = bind_default(context);
+		if (ret) {
+			return ret;
+		}
```
Same thing here, no need for the else branch, just return after calling `net_conn_update()`

sasataku replied:
Fixed

jukkar reviewed (approved):
LGTM.

@sasataku thank you for your contribution, it is much appreciated!

sasataku commented:
@jukkar
You're welcome. I'm also happy to contribute to zephyr project.
